test/parallel/test-repl-custom-eval.js

```diff
@@ -0,0 +1,132 @@
+    });
+    assert.match(
+      output,
+      /Convert this to upper case\r\n'CONVERT THIS TO UPPER CASE\\n'/
+    );
+  });
+
+  it('surfaces errors as expected', () => {
+    const output = getReplOutput('Convert this to upper case', {
+      terminal: true,
+      eval: (_code, _ctx, _replRes, cb) => cb(new Error('Testing Error')),
+    });
+    assert.match(output, /Uncaught Error: Testing Error\n/);
+  });
+
+  it('provides a repl context to the eval callback', async () => {
+    const context = await new Promise((resolve) => {
+      const r = repl.start({
+        eval: (_cmd, context) => resolve(context),
+      });
+      r.context = { foo: 'bar' };
+      r.write('\n.exit\n');
+    });
+    assert.strictEqual(context.foo, 'bar');
+  });
+
+  it('provides a global context to the eval callback', async () => {
+    const context = await new Promise((resolve) => {
+      const r = repl.start({
+        useGlobal: true,
+        eval: (_cmd, context) => resolve(context),
+      });
+      global.foo = 'global_bar';
+      r.write('\n.exit\n');
+    });
+
+    assert.strictEqual(context.foo, 'global_bar');
+    delete global.foo;
+  });
+
+  it('does not access the global context if `useGlobal` is false', async () => {
+    const context = await new Promise((resolve) => {
+      const r = repl.start({
+        useGlobal: false,
+        eval: (_cmd, context) => resolve(context),
+      });
+      global.foo = 'global_bar';
+      r.write('\n.exit\n');
+    });
+
+    assert.notStrictEqual(context.foo, 'global_bar');
+    delete global.foo;
+  });
+
+  /**
+   * Default preprocessor transforms
+   *  function f() {}  to
+   *  var f = function f() {}
+   * This test ensures that original input is preserved.
+   * Reference: https://github.com/nodejs/node/issues/9743
+   */
+  it('preserves the original input', async () => {
+    const cmd = await new Promise((resolve) => {
+      const r = repl.start({
+        eval: (cmd) => resolve(cmd),
+      });
+      r.write('function f() {}\n.exit\n');
+    });
+    assert.strictEqual(cmd, 'function f() {}\n');
+  });
+
+  it("doesn't show previews by default", () => {
+    const input = "'Hello custom' + ' eval World!'";
+    const output = getReplOutput(input, {
+      terminal: true,
+      eval: (code, _ctx, _replRes, cb) => cb(null, eval(code)),
+    }, false);
+    assert.strictEqual(output, input);
+    assert.doesNotMatch(output, /Hello custom eval World!/);
+  });
+
+  it('does show previews if `preview` is set to `true`', () => {
+    const input = "'Hello custom' + ' eval World!'";
+    const output = getReplOutput(input, {
+      terminal: true,
+      eval: (code, _ctx, _replRes, cb) => cb(null, eval(code)),
+      preview: true,
+    }, false);
+
+    const escapedInput = input.replace(/\+/g, '\\+');
+    assert.match(
+      output,
+      new RegExp(`${escapedInput}\n// 'Hello custom eval World!'`)
+    );
+  });
+});
+
+function getReplOutput(input, replOptions, run = true) {
```

Comment:
nit: consider moving this before it's being used for better readability

Reply:
I purposely put it at the end so that people ready the file not caring about the implementation details would just get the gist of it based on the function's name (while being able to scroll down to see the implementation if they wanted), that's generally my style for this sort of things

but I don't particularly mind the function being up top either if you prefer that, moved :slightly_smiling_face::+1: